File: ccpack.py
"""
This is the local component for ccpack.
You can use this to create the baked_manifest.json file needed for ccpack to work.

This requries a local file with the name ccpack.json to be present in the same directory as this file.
This file should contain a JSON object with the following structure:
{
    "name": "ccpack",
    "version": "1.0.0",
    "author": "Your Name",
}

This script will look through the current directory recursively and create a baked_manifest.json file
with all the lua files found.

If you have dependencies, add them as a submodule in your git repository.
If they have a ccpack.json file, it will be used to create the baked_manifest.json file. (the author, name and version will be used from the ccpack.json file)
"""

import os
import json


def create_baked_manifest():
    # Load the ccpack.json file
    with open('ccpack.json', 'r') as f:
        ccpack_data = json.load(f)

    # Create the baked_manifest.json structure
    baked_manifest = {
        "name": ccpack_data["name"],
        "version": ccpack_data["version"],
        "url": "https://github.com/" + ccpack_data["author"],
        "files": []
    }

    # Walk through the current directory recursively, ignore the ccpack directory itself
    current_ccpack_data = ccpack_data.copy()
    ccpack_data_relative_dir = '.'
    for root, dirs, files in os.walk('.'):
        # Ignore the ccpack directory itself
        if 'ccpack' in dirs:
            dirs.remove('ccpack')

        # ignore .git directory
        if '.git' in dirs:
            dirs.remove('.git')

        print(f"entering directory: {root}")

        # Check for ccpack.json in the current directory
        if 'ccpack.json' in files:
            with open(os.path.join(root, 'ccpack.json'), 'r') as ccpack_file:
                current_ccpack_data = json.load(ccpack_file)
            ccpack_data_relative_dir = os.path.relpath(root, start='.')
            print("Found ccpack.json, using its data: ", current_ccpack_data)

        for file in files:
            if file.endswith('.lua'):
                file_path = os.path.join(root, file)
                relative_path = os.path.relpath(file_path, start=ccpack_data_relative_dir)

                baked_file = {
                    "path": relative_path,
                    "url": "https://raw.githubusercontent.com/" + current_ccpack_data["author"] + "/" + current_ccpack_data["name"] + "/refs/heads/main/" + relative_path
                }
                print(f"Adding file to baked_manifest: {baked_file}")

                # Add the file to the baked_manifest
                baked_manifest["files"].append(baked_file)
        

    # Write the baked_manifest.json file
    with open('baked_manifest.json', 'w') as f:
        json.dump(baked_manifest, f, indent=4)

if __name__ == "__main__":
    create_baked_manifest()
    print("baked_manifest.json created successfully.")

File: ccpack.lua
--[[
    CCPack Lua script for packaging a project for use with ComputerCraft

    This script is designed to be run in a ComputerCraft environment.

    Usage:
    1. wget https://raw.githubusercontent.com/Al-Andrew/ccpack/main/ccpack.lua
    2. ccpack.lua <package_you_want_installed>

    This the package name will be something like "Al-Andrew/cc-afk" (it should be hosted on GitHub)
    This script will look for a baked_manifest.json file in the root of the repository.
    If it exists, it will use that to install the package.

    The files get installed in a directory named after the package in the current working directory.
--]]


---@class BakedFile
---@field path string The path to the file relative to the package root
---@field url string The URL to download the file from

---@class BakedManifest
---@field name string The name of the package
---@field version string The version of the package
---@field url string The URL to download the package from
---@field files table<BakedFile> A list of files to download for the package


--- Does a http get on the url looking for a baked_manifest.json file
---@param package_name string The name of the package
---@return BakedManifest The parsed manifest object
local function get_manifest(package_name)
    local url = "https://raw.githubusercontent.com/" .. package_name .. "/main/baked_manifest.json" -- todo: support other branches/versions
    -- local response = http.get(url)

    -- cant use http by default, just use wget
    shell.run("wget " .. url .. " baked_manifest.json")

    if not fs.exists("baked_manifest.json") then
        error("Failed to download manifest from " .. url)
    end

    local file = fs.open("baked_manifest.json", "r")
    if not file then
        error("Failed to open manifest file: baked_manifest.json")
    end

    ---@type string
    local content = file.readAll()
    file.close()
    
    fs.delete("baked_manifest.json") -- clean up the manifest file after reading
    
    if not content or content == "" then
        error("Manifest file is empty: baked_manifest.json")
    end

    return textutils.unserializeJSON(content)
end


--- Creates a directory if it does not exist, including parent directories
--- @param path string The path to the directory to create
local function mkdir_if_not_exists_with_parent(path)
    if fs.exists(path) then
        return
    end
    
    local parent = fs.getDir(path)
    if not fs.exists(parent) then
        mkdir_if_not_exists_with_parent(parent)
    end
    fs.makeDir(path)
end

--- Downloads a file from the given URL and saves it to the specified path
---@param path string The path to save the file to
---@param url string The URL to download the file from
local function download_file_to_path(path, url)
    mkdir_if_not_exists_with_parent(fs.getDir(path))
    
    local response = http.get(url)
    if not response then
        error("Failed to download file from " .. url)
    end

    local file = fs.open(path, "wb")
    if not file then
        error("Failed to open file for writing: " .. path)
    end

    file.write(response.readAll())
    file.close()
    response.close()
end

--- Installs a package by downloading its files according to the manifest
---@param package_name string The name of the package to install
local function install_package(package_name)
    print("Installing package: " .. package_name)

    local manifest = get_manifest(package_name)
    if not manifest or not manifest.files then
        error("Invalid manifest for package: " .. package_name)
    end
    print("Got manifest!")
    print("Package name: " .. manifest.name)
    print("Package version: " .. manifest.version)
    print("Package URL: " .. manifest.url)

    -- we put the package in a directory named after the package
    local package_dir = fs.combine(fs.getDir(shell.getRunningProgram()), manifest.name)
    mkdir_if_not_exists_with_parent(package_dir)

    print("Created working directory: " .. package_dir)

    for _, file in ipairs(manifest.files) do
        print("Downloading file: " .. file.path .. " from " .. file.url)
        download_file_to_path(fs.combine(package_dir, file.path), file.url)
    end
end


--- Driver code
local package_name = arg[1]
if not package_name then
    error("Usage: ccpack.lua <package_name>")
end

install_package(package_name)